python/lsst/daf/persistence/butler.py

```diff
@@ -275,26 +364,25 @@ def queryMetadata(self, datasetType, format=None, dataId={}, **rest):
     def datasetExists(self, datasetType, dataId={}, **rest):
         """Determines if a dataset file exists.
 
-        @param datasetType (str)   the type of dataset to inquire about.
-        @param dataId (dict)       the data id of the dataset.
-        @param **rest              keyword arguments for the data id.
+        @param datasetType (str) the type of dataset to inquire about.
+        @param dataId (DataId, dict) the data id of the dataset.
+        @param **rest keyword arguments for the data id.
         @returns (bool) True if the dataset exists or is non-file-based.
         """
 
         datasetType = self._resolveDatasetTypeAlias(datasetType)
-        dataId = copy.copy(dataId)
+        dataId = DataId(dataId)
         dataId.update(**rest)
 
-        locations = self.repository.map(datasetType, dataId)
-        if locations is None:
+        location = None
+        for repoData in self.inputs:
+            if not dataId.tag or len(dataId.tag.intersection(repoData.tags)) > 0:
+                location = repoData.repo.map(datasetType, dataId)
+                if location:
+                    break
+
+        if location is None:
             return False
-        try:
-            if len(locations) != 1:
-                raise RuntimeError("Multiple (or none) locations for datasetExists(%s, %s)" %(datasetType, dataId))
-            location = locations[0]
-        except TypeError:
-            # locations might not be a list; that's ok.
-            location = locations
 
         additionalData = location.getAdditionalData()
         storageName = location.getStorageName()
```

Comment:
Eventually I would hope that this PosixStorage-specific determination of existence should move into that class.